### DISCUSSION TITLE
Stochastic RSI

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#236] [Indicators]
Community discussion about the [Stochastic RSI](https://dotnet.stockindicators.dev/indicators/StochRsi/#content) indicator

![](https://github.com/DaveSkender/Stock.Indicators/assets/8432125/5cefdcd9-52b1-4593-a99a-dbfeffc8f9ae)

**Comment by gorazspajzer** [comment]
Hi!

I tried StohasticRSI and I compared results with StohasticRSI calculated by the stock market and found out that there are big differences. Did anyone done some comparison with exchanges?

**Reply by DaveSkender** [reply]
You'll have to provide a lot more information before anyone can analyze what you're experiencing.  Differences between platforms can be due to many different factors, including bugs in the systems you're using to compare (see #318 as an example).

**Reply by DaveSkender** [reply]
Also, keep in mind Stochastic RSI is often confused with [Stochastic Oscillator](https://daveskender.github.io/Stock.Indicators/indicators/Stoch/#content)

**Reply by gorazspajzer** [reply]
The problem was in the data volume. Now when i added more data everything is ok! Sorry for reporting this "non existing error". Keep on with the good work!

**Reply by DaveSkender** [reply]
Yes, because there are several layered smoothing techniques used here, the warmup period can be as high as `10×rsiPeriods` before you get fully converged precision.  This one is worse than most since it is using both RSI and Stochastic methods.  At a minimum, we recommend that you do not use the first `rsiPeriods+100` results.  See our [RemoveWarmupPeriods()](https://daveskender.github.io/Stock.Indicators/utilities#remove-warmup-periods) helper function -- it can do this removal for you.

**Comment by Salardx** [comment]
hi,
how can we get the point where K% and D% cross eachother?

**Reply by DaveSkender** [reply]
At some point when I get around to extended analysis (#320) I'll add a both cross and bullishness information.  For now, you'll need to compare one bar to the next to see if they reversed to check for crossover or crossunder events.  Something like:

```csharp

var r = quotes.GetStochRsi(...).ToList();

for(int i = 1; i < r.Count(); i++){
  bool isCrossover  = (r[i-1].StochRsi < r[i-1].Signal && r[i].StochRsi >= r[i].Signal);
  bool isCrossunder = (r[i-1].StochRsi > r[i-1].Signal && r[i].StochRsi <= r[i].Signal);
  DateTime date = r[i].Date;
}
```

**Reply by Salardx** [reply]
yeah, but i need to know the exact cross location, i mean time/percentage

**Reply by DaveSkender** [reply]
I added `date` to the example.  I think I know what you're seeking; however, can you elaborate on the use-case you're seeking?  I can probably give a better example.

**Reply by Salardx** [reply]
Yeah, Thank you
There's a strategy suggested to us that we enter to long position when these D% and K% cross eachother under 20% (and also short above 80%)
so we need to have the exact position and time when these two lines cross,
The example you provided shows the last time only, there's a time between the points, we need it (and also its cross percentage)

**Reply by DaveSkender** [reply]
Here's my take on what you're asking.  This is sorta a simplified backtester.  I've added this code to our [code samples](https://daveskender.github.io/Stock.Indicators/examples/#content), so you can download and run it.

```csharp
// calculate Stochastic RSI
List<StochRsiResult> resultsList =
    quotesList
    .GetStochRsi(14, 14, 3, 1)
    .ToList();

// initialize
decimal trdPrice = 0;
decimal trdQty = 0;
decimal rlzGain = 0;

Console.WriteLine("   Date         Close  StRSI Signal  Cross    Net Gains");
Console.WriteLine("-------------------------------------------------------");

// roll through history
for (int i = 1; i < quotesList.Count; i++)
{
    Quote q = quotesList[i];
    StochRsiResult e = resultsList[i];     // evaluation period
    StochRsiResult l = resultsList[i - 1]; // last (prior) period
    string cross = string.Empty;

    // unrealized gain on open trade
    decimal trdGain = trdQty * (q.Close - trdPrice);

    // check for LONG event
    // condition: Stoch RSI was <= 20 and Stoch RSI crosses over Signal
    if (l.StochRsi <= 20
     && l.StochRsi < l.Signal
     && e.StochRsi >= e.Signal
     && trdQty != 1)
    {
        // emulates BTC + BTO
        rlzGain += trdGain;
        trdQty = 1;
        trdPrice = q.Close;
        cross = "LONG";
    }

    // check for SHORT event
    // condition: Stoch RSI was >= 80 and Stoch RSI crosses under Signal
    if (l.StochRsi >= 80
     && l.StochRsi > l.Signal
     && e.StochRsi <= e.Signal
     && trdQty != -1)
    {
        // emulates STC + STO
        rlzGain += trdGain;
        trdQty = -1;
        trdPrice = q.Close;
        cross = "SHORT";
    }

    if (cross != string.Empty)
    {
        Console.WriteLine(
        $"{q.Date,10:yyyy-MM-dd} " +
        $"{q.Close,10:c2}" +
        $"{e.StochRsi,7:N1}" +
        $"{e.Signal,7:N1}" +
        $"{cross,7}" +
        $"{(rlzGain + trdGain),13:c2}");
    }
}
```

**Reply by Salardx** [reply]
Thank you, great tools, carry on!

**Comment by dehghani-mehdi** [comment]
Does this produce the same value as DTOSC? (Dynamic Trader has an indicator called DTOSC and Rober Miner uses that a lot)
